lazyClient.go

```diff
@@ -138,6 +131,26 @@ func (l *lazyClientConn) Write(b []byte) (int, error) {
 }
 
 // Close closes the underlying io.ReadWriteCloser
+//
+// This does not flush anything.
 func (l *lazyClientConn) Close() error {
+	// As the client, we flush the handshake on close to cover an
+	// interesting edge-case where the server only speaks a single protocol
+	// and responds eagerly with that protocol before waiting for out
+	// handshake.
+	//
+	// Again, we must not read the error because the other end may have
+	// closed the stream for reading. I mean, we're the initiator so that's
+	// strange... but it's still allowed
+	_ = l.Flush()
```

Comment:
Making this sync seems fine by me. I suspect we could make Close async if we wanted to (e.g. have a single background goroutine that piles up Flushes to write and times them out after a few seconds), but there seem to be plenty of other Closer implementations out there that do not have asynchronous closes (e.g. golang/x's [websocket transport](https://github.com/golang/net/blob/ff519b6c91021e6316e1df005bc19f266994ddda/websocket/websocket.go#L235)).

Do you suspect there'll be many gotchas as we (and application developers) bubble up this behavior where close is now blocking?

Reply:
It should be fine. As a matter of fact, close is already blocking for yamux and mplex.